File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.java
class Solution {
    public int solution(int[] bandage, int health, int[][] attacks) {
        int answer = 0;
       
        int maxHealth = health;       
        for(int i = 0; i < attacks.length; i++) {
            
            
            if (i == 0) {
                health -= attacks[i][1];
            }
            else {
                int time = attacks[i][0] - attacks[i-1][0] -1;
             
                if(time == 0) {
                    health -= attacks[i][1];
                    continue;
                }
                health += time * bandage[1] + (time / bandage[0]) * bandage[2];
                if (health >= maxHealth) {
                    health = maxHealth;
                }
                health -= attacks[i][1];
            }
            if(health <= 0) {
                return -1;
            }
         
            
        }
    
        return health <= 0 ? -1 : health;
    }
}

File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.js
function solution(bandage, health, attacks) {
    var answer = 0;
    const [duration, heal, bonusHeal] = bandage;
    const map = {};
    let lastTime = 0;
    for(let i = 0; i < attacks.length; i++) {
        const [attackTime, damage] = attacks[i];
         map[attackTime] = damage;
        if (i === attacks.length - 1) lastTime = attackTime;
    }
    let time = 0;
    let curHealth = health;
    for (let i = 1; i <= lastTime; i++) {
        time++;
        if (map[i]) { // 공격
            curHealth -= map[i];
            time = 0;
        } else curHealth += heal;
        if (time === duration) {
            curHealth += bonusHeal;
            time = 0;
        };
        if (curHealth > health) curHealth = health;
        if (curHealth <= 0) break;
        
    }
    
    return curHealth > 0 ? curHealth : -1;
    
}